backends/bmv2/control.cpp

```diff
@@ -0,0 +1,769 @@
+}
+
+Util::IJson* Control::convertIf(const CFG::IfNode* node, cstring prefix) {
+    (void) prefix;
+    auto result = new Util::JsonObject();
+    result->emplace("name", node->name);
+    result->emplace("id", nextId("conditionals"));
+    result->emplace_non_null("source_info", node->statement->condition->sourceInfoJsonObj());
+    auto j = conv->convert(node->statement->condition, true, false);
+    CHECK_NULL(j);
+    result->emplace("expression", j);
+    for (auto e : node->successors.edges) {
+        Util::IJson* dest = nodeName(e->endpoint);
+        cstring label = Util::toString(e->getBool());
+        label += "_next";
+        result->emplace(label, dest);
+    }
+    return result;
+}
+
+/**
+    Custom visitor to enable traversal on other blocks
+*/
+bool Control::preorder(const IR::PackageBlock *block) {
+    for (auto it : block->constantValue) {
+        if (it.second->is<IR::ControlBlock>()) {
+            visit(it.second->getNode());
+        }
+    }
+    return false;
+}
+
+bool Control::preorder(const IR::ControlBlock* block) {
+    auto bt = backend->blockTypeMap.find(block);
+    if (bt != backend->blockTypeMap.end()) {
+        LOG3(bt->second->getAnnotation("pipeline"));
+        // only generate control block marked with @pipeline
+        if (!bt->second->getAnnotation("pipeline")) {
+            return false;
+        }
+    }
+
+    const IR::P4Control* cont = block->container;
+    auto result = new Util::JsonObject();
+    result->emplace("name", cont->name);
+    result->emplace("id", nextId("control"));
+    result->emplace_non_null("source_info", cont->sourceInfoJsonObj());
+
+    auto cfg = new CFG();
+    cfg->build(cont, refMap, typeMap);
+    cfg->checkForCycles();
+
+    if (cfg->entryPoint->successors.size() == 0) {
+        result->emplace("init_table", Util::JsonValue::null);
+    } else {
+        BUG_CHECK(cfg->entryPoint->successors.size() == 1, "Expected 1 start node for %1%", cont);
+        auto start = (*(cfg->entryPoint->successors.edges.begin()))->endpoint;
+        result->emplace("init_table", nodeName(start));
+    }
+
+    auto tables = mkArrayField(result, "tables");
+    auto action_profiles = mkArrayField(result, "action_profiles");
+    auto conditionals = mkArrayField(result, "conditionals");
+
+    SharedActionSelectorCheck selector_check(refMap, typeMap);
+    block->apply(selector_check);
+
+    // Tables are created prior to the other local declarations
+    for (auto node : cfg->allNodes) {
+        if (node->is<CFG::TableNode>()) {
+            auto j = convertTable(node->to<CFG::TableNode>(), action_profiles);
+            if (::errorCount() > 0)
+                return false;
+            tables->append(j);
+        } else if (node->is<CFG::IfNode>()) {
+            auto j = convertIf(node->to<CFG::IfNode>(), cont->name);
+            if (::errorCount() > 0)
+                return false;
+            conditionals->append(j);
+        }
+    }
+
+#ifndef PSA
```

Comment:
You can temporarily have #ifdefs in the code until you debug or fix something, but don't plan to have them permanently.